src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs

```diff
@@ -0,0 +1,424 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static class OpenSsl
+    {
+        #region structures
+        [StructLayout(LayoutKind.Sequential)]
+        private struct SslContext
+        {         
+            internal IntPtr sslPtr;
+            internal IntPtr readBioPtr;
+            internal IntPtr writeBioPtr;
+            internal bool isServer;          
+        }
+        #endregion
+
+        #region internal methods
+
+        //TODO Set remote certificate options
+        internal static IntPtr AllocateSslContext(long options, SafeX509Handle certHandle, bool isServer, bool remoteCertRequired)
+        {
+            IntPtr sslContextPtr = Marshal.AllocHGlobal(Marshal.SizeOf<SslContext>());
+            SslContext sslContext = new SslContext
+            {              
+                sslPtr = IntPtr.Zero,
+                readBioPtr = IntPtr.Zero,
+                writeBioPtr = IntPtr.Zero,
+                isServer = isServer,
+            };
+
+            sslContext.isServer = isServer;
+
+            try
+            {
+                IntPtr method = GetSslMethod(isServer, options);
+
+                var contextPtr = libssl.SSL_CTX_new(method);
+                if (IntPtr.Zero == contextPtr)
+                {
+                    throw CreateSslException("Failed to allocate SSL/TLS context");
```

Comment:
If we expose them, even as the messages of inner exceptions, they still need to be localized.